dataclients/kubernetes/kube.go

```diff
@@ -901,28 +914,22 @@ func (c *Client) ingressToRoutes(items []*ingressItem) ([]*eskip.Route, error) {
 			// currently handled as mandatory
 			host := []string{"^" + strings.Replace(rule.Host, ".", "[.]", -1) + "$"}
 
-			// add extra routes from optional annotation
-			for idx, r := range extraRoutes {
-				route := *r
-				route.HostRegexps = host
-				route.Id = routeIDForCustom(i.Metadata.Namespace, i.Metadata.Name, route.Id, rule.Host, idx)
-				hostRoutes[rule.Host] = append(hostRoutes[rule.Host], &route)
-			}
-
 			// update Traffic field for each backend
 			computeBackendWeights(backendWeights, rule)
 
 			for _, prule := range rule.Http.Paths {
-				if prule.Backend.Traffic > 0 {
-					pathMode := c.pathMode
-					pathModeString := i.Metadata.Annotations[pathModeAnnotationKey]
-					if pathModeString != "" {
-						var err error
-						if pathMode, err = ParsePathMode(pathModeString); err != nil {
-							return nil, err
-						}
-					}
+				pathMode := c.getPathMode(i)
+
+				// add extra routes from optional annotation
+				for idx, r := range extraRoutes {
```

Comment:
Yes this is true, but we should consider it as fix, because it's the natural way of using ingress.
In ingress you can have in one ingress multiple hosts and path specifications and if we have special routes they should be compiled to exactly these definitions. Right now we iterate through host, not the paths, which is wrong and can lead to errors across different ingress objects easily.

If we want to create a plain route we should have a CRD instead.

Reply:
I see and I find this a valid argument. But I think this approach has quite some subtle implications: handling all the combinations of <path mode> * <custom predicates in annotations> * <predicates in the custom routes>. See also my other response below.